File: python/tests/test_soak.py
"""Round-trip soak suite, spec IMPLEMENTATION_SOAK_TESTS.md.

Two run levels share one implementation:

- CI subset (default): sweep capped at 100,000 ids per profile/variant and
  10,000 random samples. Runs inside the normal pytest run.
- Full soak (opt-in): sweep to min(1e9, capacity) per profile and 1,000,000
  random samples. Selected with BASEH_SOAK=1 (a skipUnless env guard, the
  pytest idiom named by the spec); skipped cleanly otherwise.

Environment overrides (all optional):

- BASEH_SOAK=1        run the full soak level
- BASEH_SOAK_SWEEP    override the sweep bound (smoke-testing the soak path)
- BASEH_SOAK_RANDOM   override the random sample count (spec section 4)
- BASEH_SOAK_SEED     override the random-phase seed (default 42)

Every shipped tier runs in two variants: permutation on (as shipped) and a
test-only permutation-off twin built by copying the profile dict and setting
``permutation`` to ``{"enabled": False}`` — the same twin pattern as the
repetition tests. For the expandable tier that single switch disables the
permutation across all generations, since the codec derives each generation's
Feistel key from the profile's permutation field (Baseh._feistel_key).
"""

import os
import random
import time
import unittest

from baseh import (
    BLOCKED_CODE,
    Baseh,
    BasehError,
    baseh_expandable_p_v1,
    baseh_expandable_v1,
    baseh_heavy_p_v1,
    baseh_heavy_v1,
    baseh_light_p_v1,
    baseh_light_v1,
    baseh_medium_p_v1,
    baseh_medium_v1,
    baseh_minimum_p_v1,
    baseh_minimum_v1,
)

# Fixed keyed-variant test key, spec section 2 (test keys live in tests only).
_TEST_KEY = bytes.fromhex(
    "00112233445566778899aabbccddeeff00112233445566778899aabbccddeeff"
)
_TEST_KEY_ID = "soak-test"

_SOAK = os.environ.get("BASEH_SOAK") == "1"
_CI_SWEEP_CAP = 100_000
_SEED = int(os.environ.get("BASEH_SOAK_SEED", "42"))
_RANDOM_COUNT = int(
    os.environ.get("BASEH_SOAK_RANDOM", "1000000" if _SOAK else "10000")
)
_SWEEP_OVERRIDE = os.environ.get("BASEH_SOAK_SWEEP")
_SWEEP_OVERRIDE = int(_SWEEP_OVERRIDE) if _SWEEP_OVERRIDE else None

_RANDOM_LO = 1_000_000_000
_RANDOM_HI = 100_000_000_000


def _tier_builders():
    """Every shipped tier: four fixed, their keyed -p variants, expandable."""
    return [
        ("baseh-minimum-v1", baseh_minimum_v1),
        ("baseh-minimum-p-v1", lambda: baseh_minimum_p_v1(_TEST_KEY, _TEST_KEY_ID)),
        ("baseh-light-v1", baseh_light_v1),
        ("baseh-light-p-v1", lambda: baseh_light_p_v1(_TEST_KEY, _TEST_KEY_ID)),
        ("baseh-medium-v1", baseh_medium_v1),
        ("baseh-medium-p-v1", lambda: baseh_medium_p_v1(_TEST_KEY, _TEST_KEY_ID)),
        ("baseh-heavy-v1", baseh_heavy_v1),
        ("baseh-heavy-p-v1", lambda: baseh_heavy_p_v1(_TEST_KEY, _TEST_KEY_ID)),
        ("baseh-expandable-v1", baseh_expandable_v1),
        ("baseh-expandable-p-v1", lambda: baseh_expandable_p_v1(_TEST_KEY, _TEST_KEY_ID)),
    ]


def _variants(profile):
    """Permutation on (as shipped) and the permutation-off test twin."""
    twin = dict(profile)
    twin["permutation"] = {"enabled": False}
    return [("permutation-on", profile), ("permutation-off", twin)]


def _full_sweep_bound(codec):
    """min(1e9, capacity) for fixed tiers; 1e9 for expandable, per spec."""
    if codec.profile.mode == "expandable":
        return 1_000_000_000
    return min(1_000_000_000, codec.capacity())


class _SoakBase(unittest.TestCase):
    def _fail_report(self, profile_id, variant, phase, id, code, stage, err):
        self.fail(
            f"soak failure: profile={profile_id} variant={variant} phase={phase} "
            f"seed={_SEED} id={id} code={code!r} stage={stage} error={err}"
        )

    def _report(self, profile_id, variant, phase, checked, blocked, elapsed):
        rate = int(checked / elapsed) if elapsed > 0 else checked
        print(
            f"soak: profile={profile_id} variant={variant} phase={phase} "
            f"checked={checked} blocked={blocked} elapsed={elapsed:.1f}s "
            f"throughput={rate}/s"
        )

    def _sweep(self, profile, variant, bound):
        profile_id = profile["profileId"]
        codec = Baseh(profile)
        blocked = 0
        checked = 0
        start = time.time()
        for id in range(bound):
            try:
                code = codec.encode(id)
            except BasehError as err:
                if err.code == BLOCKED_CODE:
                    # Repetition filter or blocklist: expected, not a failure.
                    blocked += 1
                    continue
                self._fail_report(profile_id, variant, "sweep", id, None, "encode", err)
            try:
                decoded = codec.decode(code)
            except BasehError as err:
                self._fail_report(profile_id, variant, "sweep", id, code, "decode", err)
            if decoded.id != id:
                self._fail_report(
                    profile_id, variant, "sweep", id, code, "roundtrip",
                    f"decoded id {decoded.id} != {id}",
                )
            checked += 1
        self._report(profile_id, variant, "sweep", checked, blocked, time.time() - start)

    def _random_phase(self, profile, variant, count):
        profile_id = profile["profileId"]
        codec = Baseh(profile)
        rng = random.Random(_SEED)
        blocked = 0
        checked = 0
        start = time.time()
        for _ in range(count):
            id = rng.randrange(_RANDOM_LO, _RANDOM_HI)
            try:
                code = codec.encode(id)
            except BasehError as err:
                if err.code == BLOCKED_CODE:
                    blocked += 1
                    continue
                self._fail_report(profile_id, variant, "random", id, None, "encode", err)
            try:
                decoded = codec.decode(code)
            except BasehError as err:
                self._fail_report(profile_id, variant, "random", id, code, "decode", err)
            if decoded.id != id:
                self._fail_report(
                    profile_id, variant, "random", id, code, "roundtrip",
                    f"decoded id {decoded.id} != {id}",
                )
            checked += 1
        self._report(profile_id, variant, "random", checked, blocked, time.time() - start)

    def _run_sweep_level(self, cap):
        for name, build in _tier_builders():
            profile = build()
            bound = _full_sweep_bound(Baseh(profile))
            if cap is not None:
                bound = min(bound, cap)
            for variant, variant_profile in _variants(profile):
                with self.subTest(profile=name, variant=variant):
                    self._sweep(variant_profile, variant, bound)

    def _run_random_level(self, count):
        print(f"soak random phase: seed={_SEED} count={count}")
        for name, build in _tier_builders():
            profile = build()
            if profile.get("mode") != "expandable":
                continue
            for variant, variant_profile in _variants(profile):
                with self.subTest(profile=name, variant=variant):
                    self._random_phase(variant_profile, variant, count)


class TestSoakSweepCISubset(_SoakBase):
    """Default CI level: 100,000-id sweep per shipped profile and variant."""

    def test_sweep_subset(self):
        self._run_sweep_level(_SWEEP_OVERRIDE if _SWEEP_OVERRIDE else _CI_SWEEP_CAP)


class TestSoakRandomCISubset(_SoakBase):
    """Default CI level: 10,000 random ids against the expandable tiers."""

    def test_random_subset(self):
        self._run_random_level(min(_RANDOM_COUNT, 10_000))


@unittest.skipUnless(_SOAK, "full soak is opt-in: set BASEH_SOAK=1")
class TestSoakFull(_SoakBase):
    """Full soak level, spec sections 3-4. Runs only with BASEH_SOAK=1."""

    def test_sweep_full(self):
        self._run_sweep_level(_SWEEP_OVERRIDE)

    def test_random_full(self):
        self._run_random_level(_RANDOM_COUNT)


if __name__ == "__main__":
    unittest.main()

File: ruby/test/test_soak.rb
# frozen_string_literal: true

require "minitest/autorun"
require "baseh"

# Cross-language round-trip soak suite (spec/IMPLEMENTATION_SOAK_TESTS.md).
#
# Every shipped tier, keyed (-p) variants included, in two variants —
# permutation ON (as shipped) and permutation OFF (test-only twin, built the
# same way the repetition tests build filter-free twins: profile.merge with
# permutation disabled; for the expandable tier the single permutation block
# applies across all generations, so disabling it covers every generation).
#
# Two run levels:
#   * CI subset (default test run): sweep capped at CI_SWEEP_CAP ids per
#     profile, CI_RANDOM_COUNT random samples.
#   * Full soak: BASEH_SOAK=1 selects the full bounds of the spec (random
#     count overridable via BASEH_SOAK_RANDOM, default 1_000_000; sweep
#     overridable via BASEH_SOAK_SWEEP for smoke runs). Without BASEH_SOAK
#     the soak-level tests skip cleanly.
class TestSoak < Minitest::Test
  SOAK_KEY = ["00112233445566778899aabbccddeeff00112233445566778899aabbccddeeff"].pack("H*")
  SOAK_KEY_ID = "soak-test"
  SEED = 42
  RANDOM_LOW = 1_000_000_000
  RANDOM_HIGH = 100_000_000_000

  CI_SWEEP_CAP = 100_000
  CI_RANDOM_COUNT = 10_000

  def soak?
    ENV["BASEH_SOAK"] == "1"
  end

  def soak_sweep_override
    ENV["BASEH_SOAK_SWEEP"]&.to_i
  end

  def soak_random_count
    (ENV["BASEH_SOAK_RANDOM"] || 1_000_000).to_i
  end

  # [profile_id, builder] for every shipped tier, plain and keyed.
  def shipped_profiles
    [
      ["baseh-minimum-v1", -> { Baseh.baseh_minimum_v1 }],
      ["baseh-minimum-p-v1", -> { Baseh.baseh_minimum_p_v1(key_bytes: SOAK_KEY, key_id: SOAK_KEY_ID) }],
      ["baseh-light-v1", -> { Baseh.baseh_light_v1 }],
      ["baseh-light-p-v1", -> { Baseh.baseh_light_p_v1(key_bytes: SOAK_KEY, key_id: SOAK_KEY_ID) }],
      ["baseh-medium-v1", -> { Baseh.baseh_medium_v1 }],
      ["baseh-medium-p-v1", -> { Baseh.baseh_medium_p_v1(key_bytes: SOAK_KEY, key_id: SOAK_KEY_ID) }],
      ["baseh-heavy-v1", -> { Baseh.baseh_heavy_v1 }],
      ["baseh-heavy-p-v1", -> { Baseh.baseh_heavy_p_v1(key_bytes: SOAK_KEY, key_id: SOAK_KEY_ID) }],
      ["baseh-expandable-v1", -> { Baseh.baseh_expandable_v1 }],
      ["baseh-expandable-p-v1", -> { Baseh.baseh_expandable_p_v1(key_bytes: SOAK_KEY, key_id: SOAK_KEY_ID) }]
    ]
  end

  # Spec section 3: sweep to min(1e9, capacity). Fixed tiers report their
  # prepared capacity; the expandable tier sweeps to 1e9.
  def sweep_bound(profile)
    codec = Baseh::Baseh.new(profile)
    if profile[:mode] == "expandable"
      1_000_000_000
    else
      [1_000_000_000, codec.capacity].min
    end
  end

  def variants(profile)
    [
      ["permutation-on", profile],
      ["permutation-off", profile.merge(permutation: { enabled: false })]
    ]
  end

  def run_sweep(cap)
    shipped_profiles.each do |name, build|
      profile = build.call
      bound = [sweep_bound(profile), cap].min
      variants(profile).each do |variant, twin|
        codec = Baseh::Baseh.new(twin)
        checked = 0
        blocked = 0
        started = Process.clock_gettime(Process::CLOCK_MONOTONIC)
        (0...bound).each do |id|
          code = begin
            codec.encode(id: id)
          rescue Baseh::BasehError => e
            raise unless e.code == "BLOCKED_CODE"

            blocked += 1
            next
          end
          decoded = codec.decode(code).id
          unless decoded == id
            flunk("soak mismatch: profile=#{name} variant=#{variant} phase=sweep " \
                  "id=#{id} code=#{code} stage=decode decoded=#{decoded}")
          end
          checked += 1
        rescue Baseh::BasehError => e
          flunk("soak error: profile=#{name} variant=#{variant} phase=sweep " \
                "id=#{id} code=#{code.inspect} stage=round-trip error=#{e.code}: #{e.message}")
        end
        elapsed = Process.clock_gettime(Process::CLOCK_MONOTONIC) - started
        rate = elapsed.positive? ? ((checked + blocked) / elapsed).round : 0
        puts(format("soak sweep profile=%s variant=%s checked=%d blocked=%d elapsed=%.2fs throughput=%d ids/s",
                    name, variant, checked, blocked, elapsed, rate))
      end
    end
  end

  def run_random(count)
    rng = Random.new(SEED)
    puts("soak random seed=#{SEED} count=#{count} range=[#{RANDOM_LOW}, #{RANDOM_HIGH})")
    shipped_profiles.select { |name, _| name.start_with?("baseh-expandable") }.each do |name, build|
      profile = build.call
      variants(profile).each do |variant, twin|
        codec = Baseh::Baseh.new(twin)
        checked = 0
        blocked = 0
        started = Process.clock_gettime(Process::CLOCK_MONOTONIC)
        count.times do
          id = rng.rand(RANDOM_LOW...RANDOM_HIGH)
          code = begin
            codec.encode(id: id)
          rescue Baseh::BasehError => e
            raise unless e.code == "BLOCKED_CODE"

            blocked += 1
            next
          end
          decoded = codec.decode(code).id
          unless decoded == id
            flunk("soak mismatch: profile=#{name} variant=#{variant} phase=random " \
                  "seed=#{SEED} id=#{id} code=#{code} stage=decode decoded=#{decoded}")
          end
          checked += 1
        rescue Baseh::BasehError => e
          flunk("soak error: profile=#{name} variant=#{variant} phase=random " \
                "seed=#{SEED} id=#{id} code=#{code.inspect} stage=round-trip error=#{e.code}: #{e.message}")
        end
        elapsed = Process.clock_gettime(Process::CLOCK_MONOTONIC) - started
        rate = elapsed.positive? ? ((checked + blocked) / elapsed).round : 0
        puts(format("soak random profile=%s variant=%s seed=%d checked=%d blocked=%d elapsed=%.2fs throughput=%d ids/s",
                    name, variant, SEED, checked, blocked, elapsed, rate))
      end
    end
  end

  # CI subset: always runs, must finish in seconds.
  def test_sweep_round_trip_ci_subset
    run_sweep(CI_SWEEP_CAP)
  end

  def test_random_round_trip_ci_subset
    run_random(CI_RANDOM_COUNT)
  end

  # Full soak: opt-in via BASEH_SOAK=1 (the minitest long-test idiom: a clean
  # skip when the env var is absent).
  def test_sweep_round_trip_soak
    skip "full soak: set BASEH_SOAK=1 to run" unless soak?

    cap = soak_sweep_override || 1_000_000_000
    run_sweep(cap)
  end

  def test_random_round_trip_soak
    skip "full soak: set BASEH_SOAK=1 to run" unless soak?

    run_random(soak_random_count)
  end
end
